Role assignment UI

allisonking opened the pull request:
Closes https://github.com/ethyca/fidesplus/issues/641

~~This branches off of multiple branches, so the diff will be unseemly until those branches are merged~~

Branches are merged, though commit history has many of the commits from https://github.com/ethyca/fides/pull/2682 because of the squashed merge + I'm not good enough at rebasing 🙃 

### Code Changes

* [ ] See the ticket

### Steps to Confirm

* [ ] _list any manual steps for reviewers to confirm the changes_

### Pre-Merge Checklist

* [ ] All CI Pipelines Succeeded
* Documentation:
  * [ ] documentation complete, [PR opened in fidesdocs](https://github.com/ethyca/fidesdocs/pulls)
  * [ ] documentation [issue created in fidesdocs](https://github.com/ethyca/fidesdocs/issues/new/choose)
* [x] Issue Requirements are Met
* [x] Relevant Follow-Up Issues Created
* [x] Update `CHANGELOG.md`

### Description Of Changes

_Write some things here about the changes and any potential caveats_

allisonking commented on clients/admin-ui/src/features/common/nav/v2/nav-config.test.ts:
```diff
@@ -57,8 +57,8 @@ describe.skip("configureNavGroups", () => {
-        ScopeRegistryEnum.CLI_OBJECTS_CREATE,
-        ScopeRegistryEnum.CLI_OBJECTS_READ,
+        ScopeRegistryEnum.SYSTEM_CREATE,
+        ScopeRegistryEnum.SYSTEM_READ,
```
I didn't follow too closely the decisions made last week, but I wonder if the latest in `main` would mean we wouldn't have to do this here? might be worth a merge to see if the tests "just pass"

eastandwestwind replied:
Agree, this is not needed after pulling in latest from Dawn's branch.

allisonking commented on clients/admin-ui/src/features/user-management/AssignSystemsTable.tsx:
```diff
@@ -66,9 +56,14 @@ export const AssignSystemsDeleteTable = () => {
+                onClick={removeSystemModal.onOpen}
                 data-testid="unassign-btn"
               />
+              <RemoveSystemModal
```
I'd prefer if we're able to use the `ConfirmationModal` in `common` just so that if we have to update our confirmation modals we can do it all in once place. but let me know if there's a good reason to keep this one as a separate component 👍

eastandwestwind replied:
Sure thing, let me see about replacing this with our `ConfirmationModal`

allisonking commented on clients/admin-ui/src/features/user-management/NewUserForm.tsx:
```diff
@@ -20,9 +23,11 @@ const NewUserForm = () => {
+    router.push(`${USER_MANAGEMENT_ROUTE}`);
   };
 
+  // @ts-ignore
```
is there a reason for this ts-ignore?

eastandwestwind replied:
nope, this can be removed!

allisonking commented on clients/admin-ui/src/features/user-management/NewUserForm.tsx:
```diff
@@ -20,9 +23,11 @@ const NewUserForm = () => {
     const b64Password = utf8ToB64(values.password);
-    return createUser({ ...values, password: b64Password });
+    createUser({ ...values, password: b64Password });
+    router.push(`${USER_MANAGEMENT_ROUTE}`);
```
I think we may still want the nice toasts that indicate the user was created successfully here. also I think we may want to redirect to the profile of this user? are we able to grab their created id from `createUser` and then build the URL off of that?

eastandwestwind replied:
Good call, I'll update to redirect to user profile, and add back the toasts for successful user creation

allisonking commented on clients/admin-ui/src/features/user-management/PasswordManagement.tsx:
```diff
@@ -1,39 +1,61 @@
+          </Restrict>
+
+          {!isNewUser && user ? (
+            <Box>
```
I'd prefer to have this delete button not inside `PasswordManagement` since it doesn't have to do with password management. can we hoist it to be a sibling?

eastandwestwind replied:
Absolutely. This makes a lot more sense to me, thanks!

allisonking commented on clients/admin-ui/src/features/user-management/RemoveSystemModal.tsx:
```diff
@@ -0,0 +1,88 @@
+  const toast = useToast();
+  const [removeUserManagedSystemTrigger] = useRemoveUserManagedSystemMutation();
+
+  const handleRemoveSystem = async (system: System, activeUserId?: string) => {
```
if you replace this with the ConfirmationModal, you could pass this func into the modal's `onConfirm`

allisonking commented on clients/admin-ui/src/features/user-management/UserForm.tsx:
```diff
@@ -84,28 +97,50 @@ const UserForm = ({
+                </Text>
+                <Box marginLeft="auto">
+                  <PasswordManagement user={user} isNewUser={isNewUser} />
+                </Box>
```
can the delete icon go here?

allisonking commented on clients/admin-ui/src/features/user-management/UserForm.tsx:
```diff
@@ -47,6 +57,7 @@ export interface Props {
 }
 
 const UserForm = ({
+  user,
```
hmm why do we want the `user` here?

eastandwestwind replied:
we need `user` in `PasswordManagement` -> `DeleteUserModal`. If we move the `DeleteUserModal` out of `PasswordManagement`, we'll still need a `user` passed in here, I believe

allisonking commented on clients/admin-ui/src/types/api/models/RolesDataMapping.ts:
```diff
@@ -0,0 +1,38 @@
+import { RoleRegistryEnum } from "~/types/api";
+
+export const ROLES = [
```
good call moving this into its own file!

allisonking commented on clients/admin-ui/src/types/api/models/ScopeRegistryEnum.ts:
```diff
@@ -60,6 +56,7 @@ export enum ScopeRegistryEnum {
   STORAGE_CREATE_OR_UPDATE = "storage:create_or_update",
   STORAGE_DELETE = "storage:delete",
   STORAGE_READ = "storage:read",
+  SYSTEM_READ = "system:read",
```
hmm was this added recently? I remember thinking it wasn't a thing

eastandwestwind replied:
It's added in Dawn's BE work here- https://github.com/ethyca/fides/pull/2714/files#diff-64b8825251ee956a60b62e71f974e359f0b540694f46dd581d2b524c2a94617bR184

allisonking reviewed (commented):
haven't tested it yet, but just leaving some code comments for now! nice work picking this up 💯

pattisdr commented:
There are these role descriptions hanging out on the right with Owner text duplicated in each of them - 

<img width="1409" alt="Screenshot 2023-03-14 at 10 20 08 AM" src="https://user-images.githubusercontent.com/9755598/225049106-fee02b8a-4f4d-4f5e-9fc6-ea1cb4e5e744.png">

pattisdr commented:
I'm assuming this is pre-existing behavior, but the backend only allows the first name and last name of a user to be edited here after it's created, not the username.  Should we disable the username from being edited here? It gives the appearance you can edit but you can't.  This could be reticketed.


![Screenshot 2023-03-14 at 11 04 45 AM](https://user-images.githubusercontent.com/9755598/225065925-a2214fea-8116-412d-aa1f-d3c438f00483.png)

eastandwestwind commented:
@pattisdr I went ahead and updated the username form field to be disabled after the user is created
<img width="829" alt="Screenshot 2023-03-14 at 5 18 45 PM" src="https://user-images.githubusercontent.com/7697292/225070135-355ea20d-6425-4d20-8fab-8b71921a4fdf.png">

pattisdr commented:
Bravo! You're addressing changes faster than I'm finding them.

pattisdr commented:
I do think this should be reticketed: Saving changes on ctl datasets when you don't have permissions shows a green popup that you "Success: Successfully modified collection" although the API request returned a 403 and no data changed.


![Screenshot 2023-03-14 at 11 22 46 AM](https://user-images.githubusercontent.com/9755598/225071405-ebecdc5b-cdb5-482a-b3fa-0a02de1a3a8a.png)

eastandwestwind commented:
@pattisdr ticketed here- https://github.com/ethyca/fides/issues/2827

pattisdr commented:
This is looking really good @allisonking and @eastandwestwind.

I've reticketed separate backend work here https://github.com/ethyca/fidesplus/issues/683

allisonking commented:
Great work, @eastandwestwind ! I went in and QA'd and fixed up a few things:

1. It said "Update Password" and "Reset password" which bothered me so I changed it to "Update password" to keep our casing consistent 😎 Also added a success toast on Update password since that wasn't there before (totally out of the scope of this work, but it was easy)
2. Refactored the delete modal logic a bit so that we don't have to pass the `user` object all the way down. I recently made a change (maybe it was in this PR...) where we store the `activeUser` (i.e. the user being edited) in redux, so we can take advantage of that to avoid some prop drilling. Also, before, after you deleted a user, you'd still be left on that user's page (which shouldn't exist anymore) so I added a redirect
3. There was a sneaky bug where if you went to a user who had a role, i.e. "owner", and then went to another user who had no roles, that user would show up as being an "owner" in the role selection page. This was probably my fault from earlier, but should be fixed now

eastandwestwind commented:
Just added a slight functionality change to User Management UI for better experience.

Before, when deleting a user on their profile page, we stayed on that page, which is misleading because you can no longer edit the user if it doesn't exist.

I'm updating such that we reroute to user management table upon user delete

eastandwestwind commented:
Side note: I've noticed some flakiness with the `Can edit a system` ->  `Can go through the edit flow` cypress test in `systems.cy.ts`. Sometimes cypress is able to find `cy.getByTestId("continue-btn").click();`, and sometimes this element doesn't exist, which causes the test to fail.

We'll soon be able to remove this line of code anyway, as it's a known bug, documented here- https://github.com/ethyca/fides/issues/2788

For now, though, re-running the test suite seems to work.

eastandwestwind commented:
Reverting to Allison's last commit, since she already covered the reroute to user management page upon user delete!
